src/writer/to_code.rs

```diff
@@ -144,26 +144,45 @@ impl ToCode for Chunk {
                 vec![s1, s2]
             }
             Name(ref name) => vec![name.clone()],
-            BoxFn { ref typ } => vec![format!("let user_data: Box<Box<{}>> = Box::new(Box::new(callback));", typ)],
-            ExternCFunc { ref name, ref parameters, ref body } => {
-                let prefix = format!(r#"unsafe extern "C" fn {}("#, name);
+            BoxFn { name: None, ref typ } => vec![
+                format!("let user_data: Box<Box<{}>> = Box::new(Box::new(callback));", typ),
+            ],
+            BoxFn { name: Some(ref name), ref typ } => vec![
+                // TODO: The name is generated based on user callback data holder. Not ideal doing
+                //       it here, this way...
+                format!("let {}_data: Box<Box<{}>> = Box::new(Box::new({}));",
+                        name,
+                        typ,
+                        &name),
+            ],
+            ExternCFunc { ref name, ref parameters, ref body, ref return_value, ref bounds } => {
+                let prefix = format!(r#"unsafe extern "C" fn {}{}("#, name, bounds);
                 let suffix = ")".to_string();
                 let params: Vec<_> = parameters.iter()
                     .flat_map(|param| param.to_code(env))
                     .collect();
-                let s = format_block_one_line(&prefix, &suffix, &params, "", ", ");
-                let mut code = format_block_smart("{", "}", &body.to_code(env), " ", " ");
+                let mut s = format_block_one_line(&prefix, &suffix, &params, "", ", ");
+                if let Some(ref return_value) = return_value {
+                    s.push_str(&format!(" -> {}", return_value));
+                }
+                s.push_str(" {");
+                let body = body.to_code(env);
+                let mut code = if body.len() == 1 {
+                    vec![format!("\t{}", body[0]), "}".to_owned()]
+                } else {
+                    format_block_smart("", "}", &body, "", "")
+                };
```

Comment:
IMHO better use `let mut code = format_block("", "}", &body.to_code(env));` as `format_block_smart` can add `}` to end of line (as was before)

Reply:
I meant that all instead all if